backend/src/order.ts

```diff
@@ -354,162 +336,252 @@ export function getAllForShow(show_id: number): Promise<IAdminOrderListItem[]> {
-  var verification_hash = md5(verification).toUpperCase();
+  let res, err: Error, conn: IConnection;
 
-  if (verification_hash !== params.RETURN_AUTHCODE) {
-    log.error('Hash verification failed!',
-      {order_id: order_id, verification_hash: verification_hash, return_authcode: params.RETURN_AUTHCODE});
-    throw 'Hash verification failed!';
+  [conn, err] = await to(db.beginTransaction());
+  if (err) {
+    log.error('Payment done failed', {error: err, order_id});
+    return reject(err);
   }
 
-  log.info('Verification ok', {order_id: order_id});
+  [res, err] = await to(db.query('select status from nk2_orders where id = :order_id', {order_id}, conn));
+  if (err) {
+    log.error('Payment done failed', {error: err, order_id});
+  }
 
-  return db.query('select status from nk2_orders where id = :order_id', {order_id: order_id})
-  .then((rows) => {
-    if (rows[0].status === 'paid') {
-      log.info('Order was already paid', {order_id: order_id});
-      return null;
+  if(res[0].status === 'paid') {
+    log.info('Order was already paid', {order_id: order_id});
+    await db.rollback(conn);
+    return resolve(null);
+  }
+
+  log.log('debug', 'Updating order payment status', {order_id});
+  [res, err] = await to(db.query('update nk2_orders set status = "paid" where id = :order_id',{order_id}, conn));
+  if (err) {
+    log.error('Updating payment status failed', {error: err, order_id});
+    await db.rollback(conn);
+    return reject(err);
+  }
+
+  [res, err] = await to(db.commit(conn));
+  if (err) {
+    log.error('Updating payment status failed', {error: err, order_id});
+    return reject(err);
     }
-    return db.query('update nk2_orders set status = "paid", payment_id = :payment_id where id = :order_id', { order_id: order_id, payment_id: params.PAID})
-    .then(() => sendTickets(order_id))
-    .catch((err) => {
-      log.error('Updating payment status failed', {error: err, order_id: order_id});
-      return Promise.reject('Updating payment status failed');
-    });
-  })
-  .then(() => {
-    return get(order_id);
-  });
 
+  log.log('debug', 'Order payment status updated', {order_id});
+
+  return resolve({order_id});
+}
+
+export async function paymentCancelled(order_id: number, req: Request): Promise<any> {
+
+  let res, err: Error, cancelResp: ICancelResponse, handler: IPayment;
+
+  log.log('debug', 'Cancelling payment', {order_id});
+
+  log.log('debug', 'Create payment handler');
+  [handler, err] = await to(getPaymentHandler(order_id));
+  if (err) {
+    return reject(err);
+  }
+
+  log.log('debug', 'Handle payment cancel callback');
+  [cancelResp, err] = await to(handler.handleCancelCallback(req));
+
+    if (err) {
+    log.error('Payment cancel handling failed', {error: err,order_id});
+    return reject(err);
+  }
+
+  log.info('Payment cancel handling succeeded', {order_id});
+
+  return await to(deleteCancelledOrder(order_id));
```

Comment:
I don't think we did it previously, but would be good to check if the order is set as PAID and in that case **not** delete the order. If the order is PAID, we've already sent the tickets to the customer, so deleting the order can be baaad.

Reply:
Might be best to check it in the deleteCancelledOrder.

Reply:
Check added now